### ISSUE TITLE
Can we drop multiple ColumnFamily support

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#4002]
**Description**

Currently we use multiple column families ~ 50. This comes with a price, since each column family has his own mem table, index and bloom filter.

We can see [here](https://github.com/zeebe-io/zeebe/issues/3988) how problematic this is.

Actually I'm not really sure whether we need the different column families at all.

> The main idea behind Column Families is that they share the write-ahead log and don't share memtables and table files. By sharing write-ahead logs we get awesome benefit of atomic writes. By separating memtables and table files, we are able to configure column families independently and delete them quickly.
> _from:_ https://github.com/facebook/rocksdb/wiki/Column-Families

Since we not really want to dynamically delete nor create new ones. Furthermore we don't want to configure them separately since usage of the different column families is totally use case dependent. That is why I see no big value in them.

I think it is still possible to use our abstraction and the related enum. If we would use only one column family we would use the Enum.value as key marker and put it as first byte in the keys. In that way the keys are still ordered and we can easily seek and access them. This would mean that we only need to change the implementation for now. We would probably gain a better memory consumption. Still we need to check how the performance would be affected by that.

**Comment by npepinpe** [comment]
Let's tackle this once we have RocksDB metrics integrated and can make an educated decision.

**Comment by Zelldon** [comment]
@korthout maybe need a new priority?

**Comment by npepinpe** [comment]
Prioritized as high as this should be considered when tackling the issue of controlling RocksDB resource usage. Let's consider this as a proposed solution when tackle the more general problem.

**Comment by korthout** [comment]
Interested to see if this will be suggested as solution by https://github.com/facebook/rocksdb/tree/master/tools/advisor (found by @npepinpe)

**Comment by korthout** [comment]
Some interesting documentation I found:

> The main idea behind Column Families is that they share the write-ahead log and don't share memtables and table files. By sharing write-ahead logs we get awesome benefit of atomic writes. By separating memtables and table files, we are able to configure column families independently and delete them quickly.
Every time a single Column Family is flushed, we create a new WAL (write-ahead log). All new writes to all Column Families go to the new WAL. However, we still can't delete the old WAL since it contains live data from other Column Families. We can delete the old WAL only when all Column Families have been flushed and all data contained in that WAL persisted in table files. This created some interesting implementation details and will create interesting tuning requirements. Make sure to tune your RocksDB such that all column families are regularly flushed. Also, take a look at Options::max_total_wal_size, which can be configured such that stale column families are automatically flushed.

https://github.com/facebook/rocksdb/wiki/Column-Families#implementation

**Comment by Zelldon** [comment]
Probably we havent configured that 🤔😬

**Comment by Zelldon** [comment]
For reference 
https://dba.stackexchange.com/questions/166159/what-is-the-point-of-column-families

https://github.com/facebook/rocksdb/wiki/RocksDB-FAQ#basic-readwrite

Q: What are column families used for?

A: The most common reasons of using column families: (1) use different compaction setting, comparators, compression types, merge operators, or compaction filters in different parts of data; (2) drop a column family to delete its data; (3) one column family to store metadata and another one to store the data.

Q: What's the difference between storing data in multiple column family and in multiple rocksdb database?

A: The main differences will be backup, atomic writes and performance of writes. The advantage of using multiple databases: database is the unit of backup or checkpoint. It's easier to copy a database to another host than a column family. Advantages of using multiple column families: (1) write batches are atomic across multiple column families on one database. You can't achieve this using multiple RocksDB databases. (2) If you issue sync writes to WAL, too many databases may hurt the performance.

Q: I have different key spaces. Should I separate them by prefixes, or use different column families?

A: If each key space is reasonably large, it's a good idea to put them in different column families. If it can be small, then you should consider to pack multiple key spaces into one column family, to avoid the trouble of maintaining too many column families.

**Comment by Zelldon** [comment]
Would like to try as a spike try to completely remove the column families. Then we can see how it feels and works.

Might be that we in the end only remove some column families and merge them together. But I think if we have multiple we should try to configure them properly based on the expected load and usage.

**Comment by npepinpe** [comment]
I could imagine we can keep an API where they seem separated but underneath point to the same column family. I can also see that there might be a point where we have different settings for different column families, but we could start with everything in the same column, and migrate it when we have solid proof that we need different settings for a particular state.

**Comment by Zelldon** [comment]
This was my proposal in this issue yes:

> I think it is still possible to use our abstraction and the related enum. If we would use only one column family we would use the Enum.value as key marker and put it as first byte in the keys. In that way the keys are still ordered and we can easily seek and access them. This would mean that we only need to change the implementation for now. We would probably gain a better memory consumption. Still we need to check how the performance would be affected by that.

**Comment by Zelldon** [comment]
@npepinpe do we want to try that out? Should I do a spike?

**Comment by Zelldon** [comment]
# One column Family

I did a "spike" and tried this out on a new branch https://github.com/zeebe-io/zeebe/tree/zell-one-column-family.

## What I did

 * On opening the db we just creating one column family, which is default.
 * On accessing a column family we using a prefix which is equal to the ordinal of the enum.
 * All keys are replaced by compositekeys internally - DbLong + actualKey, where DbLong contains the ordinal
 * The prefix is written before the actual key and this is also used for seeking/iterating.
 * After some test I also migrate the prefix to direct buffers, which should it make it more performant than using byte arrays

## Results

I did for each config two benchmarks. One to test throughput, which means I have 12 workers and a starter which start 300 workflow instances per second. Another one to test latency, which means 1 worker and a starter which start 10 workflow instances per second.

### Base 

As we know with out base benchmark we use a lot of memory, especially in RocksDB around 2 gigs per partition.

#### Throughput

![base-throughput-general](https://user-images.githubusercontent.com/2758593/96153248-c63fe600-0f0d-11eb-87d9-b199ed3c7d58.png)
![base-throughput-io](https://user-images.githubusercontent.com/2758593/96153254-c7711300-0f0d-11eb-8fb8-41e2e29367b8.png)
![base-throughput-latency](https://user-images.githubusercontent.com/2758593/96153255-c7711300-0f0d-11eb-81b3-d6129e18a9fb.png)
![base-throughput-res](https://user-images.githubusercontent.com/2758593/96153256-c809a980-0f0d-11eb-8b5b-a5dc4ddc47ee.png)
![base-throughput-rocks](https://user-images.githubusercontent.com/2758593/96153258-c809a980-0f0d-11eb-987e-dbffb40389ee.png)
![base-throughput-rocks2](https://user-images.githubusercontent.com/2758593/96153260-c8a24000-0f0d-11eb-856a-afb0d0fa8a90.png)
![base-throughput-rocks3](https://user-images.githubusercontent.com/2758593/96153261-c93ad680-0f0d-11eb-9926-6911fa755bbd.png)

#### Latency

![base-latency-general](https://user-images.githubusercontent.com/2758593/96153333-dd7ed380-0f0d-11eb-8a51-4b4a195c7ad6.png)
![base-latency-io](https://user-images.githubusercontent.com/2758593/96153335-de176a00-0f0d-11eb-9be0-d7d08d9fcc41.png)
![base-latency-latency](https://user-images.githubusercontent.com/2758593/96153337-de176a00-0f0d-11eb-9c73-5a3f37d73efd.png)
![base-latency-res](https://user-images.githubusercontent.com/2758593/96153338-deb00080-0f0d-11eb-844c-3e260eb9f76e.png)
![base-latency-rocks](https://user-images.githubusercontent.com/2758593/96153339-deb00080-0f0d-11eb-89ef-f5ade6acef16.png)
![base-latency-rocks2](https://user-images.githubusercontent.com/2758593/96153341-df489700-0f0d-11eb-9717-490785e85dfd.png)
![base-latency-rocks3](https://user-images.githubusercontent.com/2758593/96153343-dfe12d80-0f0d-11eb-9162-34ef8ac20971.png)

### One Column Family

#### Throughput
In the throughput benchmark we see a major difference, a factor 3 degrade. **BUT** we can also see that we using much less memory in RocksDB actually by factor 10. We using per partition now only ~200 MB.

![one-col-throughput-general](https://user-images.githubusercontent.com/2758593/96153547-1d45bb00-0f0e-11eb-87d2-400ef611b544.png)
![one-col-throughput-io](https://user-images.githubusercontent.com/2758593/96153552-1dde5180-0f0e-11eb-9aaf-db9981cb0c83.png)
![one-col-throughput-latency](https://user-images.githubusercontent.com/2758593/96153554-1e76e800-0f0e-11eb-92b9-e4ed59c9ec33.png)
![one-col-throughput-res](https://user-images.githubusercontent.com/2758593/96153556-1e76e800-0f0e-11eb-843a-514ac4e29bdd.png)
![one-col-throughput-rocks](https://user-images.githubusercontent.com/2758593/96153560-1f0f7e80-0f0e-11eb-9f93-ad47a52c2191.png)
![one-col-throughput-rocks2](https://user-images.githubusercontent.com/2758593/96153562-1f0f7e80-0f0e-11eb-81e4-5af782555be9.png)
![one-col-throughput-rocks3](https://user-images.githubusercontent.com/2758593/96153564-1fa81500-0f0e-11eb-8c89-36b28ad1aa5d.png)

#### Latency

The latency looks actually similar to our base latency test.

![one-col-latency-general](https://user-images.githubusercontent.com/2758593/96153580-23d43280-0f0e-11eb-82cc-a2dd89a35db2.png)
![one-col-latency-io](https://user-images.githubusercontent.com/2758593/96153582-246cc900-0f0e-11eb-8896-e35461f26206.png)
![one-col-latency-latency](https://user-images.githubusercontent.com/2758593/96153583-246cc900-0f0e-11eb-83fa-535c5cbe58f6.png)
![one-col-latency-res](https://user-images.githubusercontent.com/2758593/96153584-25055f80-0f0e-11eb-926f-70459c2fa4ca.png)
![one-col-latency-rocks](https://user-images.githubusercontent.com/2758593/96153587-25055f80-0f0e-11eb-92f4-69c9ae246526.png)
![one-col-latency-rocks2](https://user-images.githubusercontent.com/2758593/96153590-259df600-0f0e-11eb-812d-11cded339240.png)
![one-col-latency-rocks3](https://user-images.githubusercontent.com/2758593/96153591-259df600-0f0e-11eb-8154-dbe3664ecb65.png)


### One Column Family Direct Buffers

Since we know that using byte arrays is a bit problematic with JNI etc. I changed the prefix search and column key usage to direct buffers. Since before we saw no real difference in latency I just tested throughput.

#### Throughput

It looks even a bit worse then before, but this might be also related to leader distributions.

![one-col-throughput-direct-general](https://user-images.githubusercontent.com/2758593/96154154-ce4c5580-0f0e-11eb-89ec-1d6d27ccf355.png)
![one-col-throughput-direct-io](https://user-images.githubusercontent.com/2758593/96154159-cee4ec00-0f0e-11eb-839a-19252a638c55.png)
![one-col-throughput-direct-latency](https://user-images.githubusercontent.com/2758593/96154162-cf7d8280-0f0e-11eb-8cea-57ed28957256.png)
![one-col-throughput-direct-res](https://user-images.githubusercontent.com/2758593/96154163-d0161900-0f0e-11eb-86fd-122a00dc4b6c.png)
![one-col-throughput-direct-rocks](https://user-images.githubusercontent.com/2758593/96154167-d0161900-0f0e-11eb-8196-d1e199da6307.png)
![one-col-throughput-direct-rocks2](https://user-images.githubusercontent.com/2758593/96154168-d0aeaf80-0f0e-11eb-9f44-d4bb054c9684.png)
![one-col-throughput-direct-rocks3](https://user-images.githubusercontent.com/2758593/96154169-d0aeaf80-0f0e-11eb-9c29-afd6f0805abf.png)

### Further Investigation

I investigated a bit why the throughput might be so bad and I saw that the most of the time the cpu is doing `RocksDbIterator#seek0`, espcially because of `whileEqualPrefix`.

![flame](https://user-images.githubusercontent.com/2758593/96154457-1cf9ef80-0f0f-11eb-94c0-5ee7ebbf663a.png)

I discussed this with @deepthidevaki @korthout. @korthout mentioned that this is can be related to using one column family and now having the complete data in one LSM tree and that searching for a specific prefix can take now a while, because it has to search for several levels before finding the right one. Which means seeking now takes longer, BUT we actually using our resources better then before :) 

**Note**: Personally I would like to have something like an Red-Black Tree, where you can access faster the prefix and then search in the correct sst file or what ever, but this is currently not how it works in RocksDB.

@deepthidevaki and @korthout  mentioned that it might be possible to configure the Bloom filters and properly also other stuff like size of the memtable to get better performance results. 

@npepinpe question to you should we spent more time in investigating this? What do you think? Personally I liked that we could be able to reduce our footprint a lot not only memory also disk space usage.

**Comment by npepinpe** [comment]
Let's not spend time on this any more except in the context of @deepthidevaki's current work investigating the memory usage of Zeebe (if it seems relevant to that, of course). I will downgrade the priority here, but keep it open because it may still be a valid solution to our memory issues - this will depend on the findings from @deepthidevaki.

**Comment by npepinpe** [comment]
Deepthi had an interesting idea, which is we could aggregate data into a small subset of columns based on access pattern. That still needs some research into what's the best way to aggregate - key/value size, iteration pattern, write/read workload, etc. - but I think it's a nice lead to find a compromise between many column families and a single one.

I'm not sure yet how to prioritize this however, I could use some brainstorming powers. The main reasons to investigate a different approach, from what I can see, are smaller memory footprint, and ease of controlling the resource usage. The drawback with the 1-column solution was performance - so we'd need to find a good middle ground, where performance is adequate, and memory footprint is smaller/easier to control/predict. Are there any other potential advantages? Memory issues don't plague us too much in our prod environment since we run on preemptible nodes - however we face issues when dealing with highly fragmented snapshots (replication/checksum computation is expensive due to high number of files). Would having less column families help? Are there other issues it would help us with?

Would it help better if we listed the issues we face with RocksDB, then create specific issues to tackle these? Or do we expect they have a similar underlying cause, i.e. too many column families? 

/cc @korthout @deepthidevaki @Zelldon since you've all spent time with RocksDB :slightly_smiling_face:

**Comment by npepinpe** [comment]
Let's start of the year on a good note :tada: One column family only with better performance than our normal configuration. Also this would've been impossible with @Zelldon's work and help :muscle: 

You can see the branch [here](https://github.com/zeebe-io/zeebe/compare/np-1cf). You will find almost all configuration options for RocksDB commented, hopefully it's clear, but please challenge them. I will paste the code blocks at the end of this comment as well.

You can also check out the different load tests I ran (with different memory, number of memtables, etc.) - np-1cf, np-1cf-2, np-1cf-3, np-1cf-4, and on the long running cluster, np-1cf-long (`np-1cf-4` and `np-1cf-long` both have the latest configuration).

All the load tests are ran with our standard load tests (1 service task, etc., 3 nodes, 3 partitions, replication factor 3). The only things changed was using a single column family with prefixes, and RocksDB settings.

A quick reminder of our baseline: we reach between 120-150 wfi/s, with an average latency between 2 and 5 seconds.

TL;DR - when leaders are evenly distributed, with 640Mb allocated per RocksDB instances (and yes it stays around this limit as far as I can tell), it achieves 300 wfi/s with an average latency between 1 and 2 seconds. When one node is leader for 2, and one for 1, it drops to ~270 wfi/s, latency remains similar. When one node is leader for all, it drops again to ~230-250 wfi/s, latency remains the same. Snapshots have generally few files, memory usage seems to be more or less bounded by what we budget for RocksDB.

Some of the screenshots I took from `np-1cf-3` as it's been running for over 24h, whereas `np-1cf-4` has only been running for an hour or so.

![Screenshot from 2021-01-01 11-52-41](https://user-images.githubusercontent.com/43373/103437626-0029aa00-4c2a-11eb-8a71-d5dbfd94fe94.png)
![Screenshot from 2021-01-01 12-09-17](https://user-images.githubusercontent.com/43373/103437654-349d6600-4c2a-11eb-9a8a-8cc383ae4392.png)
![Screenshot from 2021-01-01 11-54-27](https://user-images.githubusercontent.com/43373/103437624-fef87d00-4c29-11eb-9457-ffc4cc91656b.png)
![Screenshot from 2021-01-01 11-53-24](https://user-images.githubusercontent.com/43373/103437625-ff911380-4c29-11eb-82ad-e92cf8088ae9.png)
![Screenshot from 2021-01-01 11-55-47](https://user-images.githubusercontent.com/43373/103437623-fe5fe680-4c29-11eb-8213-74ecb00376b6.png)
![Screenshot from 2021-01-01 11-56-50](https://user-images.githubusercontent.com/43373/103437622-fd2eb980-4c29-11eb-8ec1-d79cd230b304.png)


Let's go into more details about the changes then -

The first, simplest change, was disabling the WAL. Normally every write in RocksDB goes concurrently into a memtable and a WAL. This allows RocksDB to be recoverable in the case of an unexpected crash. However, we only write to our `runtime` RocksDB, which we always throw away on crash, so the WAL provides us nothing.

The first is only the default column family is used. This is done by adding a 2 bytes prefix to every key which is simply the ordinal of the enum value. This should be changed to have the enum implement a specific interface which should provide the value explicitly - this will allow us to be backwards and forwards compatible. It was pretty easy to hide this from the consumers since we perform all accesses via interfaces already (i.e. `ColumnFamily`), so the prefix can be added in the implementation at runtime by taking in the `KeyType key` the consumer passes, and giving to RocksDB a `DbCompositeKey(prefix, key)`.

> One of the reason for re-exploring this is that there's an overhead to column families, not just in terms of CPU and memory, but also in terms of disk usage. [As per the documentation](https://github.com/facebook/rocksdb/wiki/Column-Families#implementation), column families only share a WAL - they each have their own LSM tree, their own memtables, indexes, filters, readers, etc. As we don't use the WAL, we gain little from using multiple column families other than slightly faster out of the box performance when it comes to seek (incidentally because each seek has to lookup less data). 

Without any configuration changes, as @Zelldon mentioned, performance is then abysmal. However, by having a single column family with prefixes, our workload becomes a mix of writes (including deletes, but mostly puts) and point lookups. While we do iterate, the most expensive operations we now run are the `Iterator#seek` and the `DB#get`. To optimize these, we can define a `prefix_extractor` using the `useFixedLengthPrefixExtractor(Short.BYTES)`, and set a positive `setMemtablePrefixBloomSizeRatio`. This will create an additional index for the prefixes which will greatly speed up point lookup for our "column" prefixes. Other configurations improve this use case, namely using `kHashSearch` for the index type (as opposed to binary search), and `kDataBlockBinaryAndHash` for the data index type (with an appropriate ratio - see [this blog post](https://rocksdb.org/blog/2018/08/23/data-block-hash-index.html)).

Then comes the memory budgeting - one of the initial goal was to simplify how we provision memory for RocksDB. This becomes a lot simpler with a single column family. The configuration takes a single total memory budget, allocates a third for the block cache (a RocksDB recommendation - see [here](https://github.com/facebook/rocksdb/wiki/Setup-Options-and-Basic-Tuning#block-cache-size)), and the rest to the memtables. Bloom filters and indexes are accounted for by putting them in the cache as pinned resources.

There were many other settings added which seemed to improve, though it's unclear if they're optimal or not - increasing the block size reduced the memory usage incurred by the block cache, at the cost of potentially more I/O (but from the perf dumps it looks OK). Compression was disabled for L1 and L2 (L0 is always uncompressed), as the data is not so big there, so there isn't much use for it. 

Here are the configuration blocks:

```java
final DBOptions dbOptions =
          new DBOptions()
              .setErrorIfExists(false)
              .setCreateIfMissing(true)
              .setParanoidChecks(true)
              // 1 flush, 1 compaction
              .setMaxBackgroundJobs(2)
              // we only use the default CF
              .setCreateMissingColumnFamilies(false)
              // may not be necessary when WAL is disabled, but nevertheless recommended to avoid
              // many small SST files
              .setAvoidFlushDuringRecovery(true)
              // fsync is called asynchronously once we have at least 4Mb
              .setBytesPerSync(4 * 1024 * 1024L)
              // limit the size of the manifest (logs all operations), otherwise it will grow
              // unbounded
              .setMaxManifestFileSize(256 * 1024 * 1024L)
              // speeds up opening the DB
              .setSkipStatsUpdateOnDbOpen(true)
              // can be disabled when not profiling
              .setStatsDumpPeriodSec(20)
              .setStatistics(statistics);
```

```java
// given
    final long totalMemoryBudget = 512 * 1024 * 1024L; // make this configurable
    // recommended by RocksDB, but we could tweak it; keep in mind we're also caching the indexes
    // and filters into the block cache, so we don't need to account for more memory there
    final long blockCacheMemory = totalMemoryBudget / 3;
    // flushing the memtables is done asynchronously, so there may be multiple memtables in memory,
    // although only a single one is writable. once we have too many memtables, writes will stop.
    // since prefix iteration is our bread n butter, we will build an additional filter for each
    // memtable which takes a bit of memory which must be accounted for from the memtable's memory
    final int maxConcurrentMemtableCount = 6;
    final double memtablePrefixFilterMemory = 0.15;
    final long memtableMemory =
        Math.round(
            ((totalMemoryBudget - blockCacheMemory) / (double) maxConcurrentMemtableCount)
                * (1 - memtablePrefixFilterMemory));

    // start with some defaults
    final var columnFamilyOptionProps = new Properties();
    // look for cf_options.h to find available keys
    // look for options_helper.cc to find available values
    columnFamilyOptionProps.put("compaction_pri", "kOldestSmallestSeqFirst");

    // apply custom options
    columnFamilyOptionProps.putAll(userProvidedColumnFamilyOptions);

    final var columnFamilyOptions =
        ColumnFamilyOptions.getColumnFamilyOptionsFromProps(columnFamilyOptionProps);
    if (columnFamilyOptions == null) {
      throw new IllegalStateException(
          String.format(
              "Expected to create column family options for RocksDB, "
                  + "but one or many values are undefined in the context of RocksDB "
                  + "[Compiled ColumnFamilyOptions: %s; User-provided ColumnFamilyOptions: %s]. "
                  + "See RocksDB's cf_options.h and options_helper.cc for available keys and values.",
              columnFamilyOptionProps, userProvidedColumnFamilyOptions));
    }

    // you can use the perf context to check if we're often blocked on the block cache mutex, in
    // which case we want to increase the number of shards (shard count == 2^shardBits)
    final Cache cache = new LRUCache(blockCacheMemory, 8, false, 0.15);
    closeables.add(cache);

    final Filter filter = new BloomFilter(10, false);
    closeables.add(filter);

    final TableFormatConfig tableConfig =
        new BlockBasedTableConfig()
            .setBlockCache(cache)
            .setFilterPolicy(filter)
            // increasing block size means reducing memory usage, but increasing read iops
            .setBlockSize(32 * 1024L)
            // caching and pinning indexes and filters is important to keep reads/seeks fast when we
            // have many memtables, and pinning them ensures they are never evicted from the block
            // cache
            .setCacheIndexAndFilterBlocks(true)
            .setPinL0FilterAndIndexBlocksInCache(true)
            .setCacheIndexAndFilterBlocksWithHighPriority(true)
            // default is binary search, but all of our scans are prefix based which is a good use
            // case for efficient hashing
            .setIndexType(IndexType.kHashSearch)
            .setDataBlockIndexType(DataBlockIndexType.kDataBlockBinaryAndHash)
            .setDataBlockHashTableUtilRatio(0.5);

    // TODO: allow settings to be overwritable
    return columnFamilyOptions
        // prefix seek must be fast, so allocate an extra 10% of a single memtable budget to create
        // a filter for each memtable, allowing us to skip them if possible
        .useFixedLengthPrefixExtractor(Short.BYTES)
        .setMemtablePrefixBloomSizeRatio(memtablePrefixFilterMemory)
        // memtables
        // merge at least 3 memtables per L0 file, otherwise all memtables are flushed as individual
        // files
        .setMinWriteBufferNumberToMerge(Math.min(3, maxConcurrentMemtableCount))
        .setMaxWriteBufferNumberToMaintain(maxConcurrentMemtableCount)
        .setMaxWriteBufferNumber(maxConcurrentMemtableCount)
        .setWriteBufferSize(memtableMemory)
        // compaction
        .setLevelCompactionDynamicLevelBytes(true)
        .setCompactionPriority(CompactionPriority.OldestSmallestSeqFirst)
        .setCompactionStyle(CompactionStyle.LEVEL)
        // L-0 means immediately flushed memtables
        .setLevel0FileNumCompactionTrigger(maxConcurrentMemtableCount)
        .setLevel0SlowdownWritesTrigger(
            maxConcurrentMemtableCount + (maxConcurrentMemtableCount / 2))
        .setLevel0StopWritesTrigger(maxConcurrentMemtableCount * 2)
        // configure 4 levels: L1 = 32mb, L2 = 320mb, L3 = 3.2Gb, L4 >= 3.2Gb
        // level 1 and 2 are uncompressed, level 3 and above are compressed using a CPU-cheap
        // compression algo. compressed blocks are stored in the OS page cache, and uncompressed in
        // the LRUCache created above. note L0 is always uncompressed
        .setNumLevels(4)
        .setMaxBytesForLevelBase(32 * 1024 * 1024L)
        .setMaxBytesForLevelMultiplier(10)
        .setCompressionPerLevel(
            List.of(
                CompressionType.NO_COMPRESSION,
                CompressionType.NO_COMPRESSION,
                CompressionType.LZ4_COMPRESSION,
                CompressionType.LZ4_COMPRESSION))
        // defines the desired SST file size (but not guaranteed, it is usually lower)
        // L0 => 8Mb, L1 => 16Mb, L2 => 32Mb, L3 => 64Mb
        // as levels get bigger, we want to have a good balance between the number of files and the
        // individual file sizes
        .setTargetFileSizeBase(8 * 1024 * 1024L)
        .setTargetFileSizeMultiplier(2)
        // misc
        .setTableFormatConfig(tableConfig);
```

There's more work to be done - first, as you can see, most of these configs are hardcoded and will overwrite what the user passes. So we'd need to make this somehow overwritable by the user's config, and also provide a nicer way to configure the memory budget for RocksDB. That's the first step. Second step is clean up the prefix stuff - it's pretty hacky now, including the whole instrumentation stuff. Finally, it would be tweaking the configuration - right now these work well, but it's unclear if they're optimal - is it better to have fewer, bigger memtables, or many smaller memtables (from basic testing the former seems better, but hard to say)? Does this really solve the fragmentation issue of snapshots, or just delays it (it should, because before SST files were generated from the WAL and from seldom used columns, whereas now everything is often touched, resulting in a higher compaction rate).

We'd probably also want to disable statistics but allow this to be turned on at times for profiling.

Sources used to get all these things:

- https://github.com/facebook/rocksdb/wiki/Iterator-Implementation (esp. how memtables and L0 are iterated)
- https://github.com/facebook/rocksdb/wiki/Prefix-Seek
- https://www.cockroachlabs.com/blog/cockroachdb-on-rocksd/
- https://github.com/juxt/crux/issues/97 and https://github.com/juxt/crux/issues/125
- https://docs.pingcap.com/tidb/dev/rocksdb-overview
- https://github.com/facebook/rocksdb/wiki/MemTable explaining the different memtables and their use cases
- https://github.com/facebook/rocksdb/wiki/Block-Cache (esp. the LRUCache section)
- https://rocksdb.org/blog/2018/08/23/data-block-hash-index.html
- https://github.com/facebook/rocksdb/wiki/Partitioned-Index-Filters
- https://github.com/facebook/rocksdb/wiki/Transactions explaining how transactions work under the hood
- https://github.com/facebook/rocksdb/wiki/RocksDB-Tuning-Guide and the examples, esp. the prefix DB one
- https://github.com/facebook/rocksdb/wiki/Column-Families
- https://github.com/facebook/rocksdb/wiki/Setup-Options-and-Basic-Tuning
- https://github.com/facebook/mysql-5.6/wiki/Schema-Design - MyRocks, a RocksDB based InnoDB alternative

**Comment by npepinpe** [comment]
Looking at the long running benchmark from the last 2 days (ignore that zeebe 1 is missing, it never started because the node pool is full - oops!)

![Screenshot from 2021-01-02 17-17-44](https://user-images.githubusercontent.com/43373/103461391-760f3d80-4d1e-11eb-9c66-f9c04344cced.png)

As you can see, memory is fairly stable - note that here Zeebe 2 is leader for 2 partitions (and has been since the beginning! surprisingly no role changes occurred), and Zeebe 0 leader for 1.

If you check the smap output on Zeebe 2, you get:

```
Rss:            13000348 kB
Pss:            12067853 kB
Pss_Anon:         555992 kB
Pss_File:       11511861 kB
Pss_Shmem:             0 kB
Shared_Clean:    1398560 kB
Shared_Dirty:          0 kB
Private_Clean:  11045780 kB
Private_Dirty:    556008 kB
Referenced:     13000348 kB
Anonymous:        555992 kB
LazyFree:              0 kB
AnonHugePages:         0 kB
ShmemPmdMapped:        0 kB
Shared_Hugetlb:        0 kB
Private_Hugetlb:       0 kB
Swap:                  0 kB
SwapPss:               0 kB
Locked:                0 kB
```

From this, you can see almost all of our memory is from mapped files (i.e. `Pss_File`), so it's mostly the OS page cache, and what we actually allocated via malloc and such is 555992 kB (about ~500 mB). The OOM killer deals very well with this as it will always force the page cache to drop pages first before killing a process, and looking at the WSS graph above we see this happening regularly.

However, what this means for us is when we want to explain how much memory to provision, we need to talk about additional things:

- memory to provision per partition for the broker's long living allocations (e.g. dispatcher)
- JVM heap memory
- RocksDB block cache and memtable allocations
- some % of memory for the OS page cache to allow both RocksDB and our log readers to work efficiently

That last one is something we need to figure out - how much memory should be provisioned that we won't use directly but will be available for buffered I/O?

So for example, in this load test, we budgeted 512Mb for RocksDB per partition for cache, memtable, index, filter, etc. - everything but temporary overhead during flushing and compaction (see for example the Memtable wiki page, which shows that `SkipList` memtable requires constant extra memory for flushing). That means for Zeebe 2, we have about 1GB for sure, and let's say some constant overhead, let's say 25%, ~256Mb - honestly I have no idea, but let's worst case this. Then we have so 25% of the requested memory for JVM heap (so 3GB - though we are using only 1GB so far), and at peak we have something like 17 segments (seems we export too slow now), which all have a writer which maps our segment files entirely (512MB), which theoretically have been read such that their pages will be loaded in the page cache and only evicted under memory pressure. We can see the WSS regularly goes up to 11GB (out of 12GB requested), but gets dropped back down rather hard, which would corroborate the hypothesis that most of the memory is page cache - which is also corroborated by the smap output itself (since only about ~500MB are anonymous allocations, the rest are all mapped from files).

Interesting would be to run a load test with less memory available - here we see a leader of 2 partitions with a stable RSS of 4GB. So let's say 2GB per partition, we can try running with 6GB and see how stable it remains under constant load for a long period of time.

**Comment by npepinpe** [comment]
On the low load benchmark (10 wfi/s), after almost 2 days, the snapshot looks like this:

```
root@np-1cf-low-zeebe-0:/usr/local/zeebe# ls -lash data/raft-partition/partitions/1/snapshots/51333299-6-1609669934641-57819686-57817749/
total 608K
4.0K drwxr-xr-x 2 root root 4.0K Jan  3 10:32 .
4.0K drwxr-xr-x 3 root root 4.0K Jan  3 10:32 ..
 24K -rw-r--r-- 1 root root  21K Jan  3 10:27 000632.sst
552K -rw-r--r-- 1 root root 551K Jan  3 10:32 000633.sst
4.0K -rw-r--r-- 1 root root   16 Jan  3 10:32 CURRENT
 12K -rw-r--r-- 1 root root  12K Jan  3 10:32 MANIFEST-000537
8.0K -rw-r--r-- 1 root root 5.9K Jan  3 10:32 OPTIONS-000535
```

**Comment by Zelldon** [comment]
Awesome 🥳😍 thanks for the further investigation looks very promising !

**Comment by npepinpe** [comment]
By the way, the performance drops quite a bit when a node is leader for all - it does around 600-700 wfi/s, whereas when it's leader for 2 it does 800-900 wfi/s, and when leader for 1 ~900 wfi/s. Didn't look much into it though, also not sure if we need to at the moment. Just a note that it's not linear.

**Comment by korthout** [comment]
This is really cool stuff. Great job!

**Comment by npepinpe** [comment]
The low load benchmark is still with only 2 files in its snapshot :+1: 

High load one however has already 34, most of which are around 8MB (the target file size). We could have fewer files by having a target file size multiplier 2, such that L0 and L1 files would be 8MB, L2 would be say 16MB, L3 32MB, and L4 64MB. But keep in mind that when replicating, we load the whole file in memory, so those 64MB files become non negligible. Possibly it's still worth it over 8 files of 8MB? :man_shrugging: 

I would like to have chunked replication instead, but that won't happen now - so maybe consider also setting a multiplier of 2 for the target file size. Just a thought.

**Comment by Zelldon** [comment]
As dicussed in the kick off we will go with the one column family and the configurations proposed by @npepinpe as first iteration.
I will today clean up and start with some benchmarks.


**Goals:**
  * We want to get RocksDB under control, memory wise.
  * We want to avoid further incidents because of too many open/created files
  * We want to have **similar** or better performance, if possible

**Nice to haves:**

  * Minimal configuration burden. Not 100 configuration parameters
      * For configuring memory usage we could have one config and calculate everything based on this
   * We want to understand better how it is actually working and share it with the team.
 
**Don’ts:**
  * We don’t want to get the best performance out of it. So not necessarily tweak the last milliseconds out of it


## What was decided:

 * we use one column family
 * **Configuration:**
    * When user sets a RocksDB cfg it overwrites everything. We should document that and put a WARNING 
     * Put it into experimental cfg
     * Statistics should be a optional feature, which is disabled by default
     * State memory size configuration in experimental. We calculate based on this value the cache, write buffer etc.
     * _Compression_ don’t worry about it now

* **Metrics:**
  * Skip for now
* DirectBuffer, lets not do it now
* Verify in separate issue:
    * disabling WAL; snapshots contain WAL
    * Verify that we don’t lose updates on snapshotting, Are memtable’s flushed on checkpointing?
* **How to verify that we reached our goals? *Acceptance Criterias kind of***
   * Low load benchmark
   * Frequent snapshots to reproduce memory issues. (on stable nodes) Run it for a week
   * Big state, see how it behaves
       * 3 million messages
       * A lot of workflow instances
       *  Timers, a lot of them

**Comment by rusterholz** [comment]
> Frequent snapshots to reproduce memory issues. (on stable nodes) Run it for a week

Just my two cents based on our experience running Zeebe in production: you may need to run a broker for 2 weeks or more to see some of the issues that we've encountered. For example, last week we had a Zeebe outage that included total data loss; one of the brokers had been running for 17 days. 

I would be interested to know: what is the longest period of time that the Zeebe team has been able to run a cluster for, under load, without encountering any errors or OOM events?

**Comment by npepinpe** [comment]
We run all our clusters on GKE preemptible nodes, which are much cheaper. This means each node gets restarted at least once a day by the system. You can read more about it [here](https://cloud.google.com/kubernetes-engine/docs/how-to/preemptible-vms).

This works for us since Zeebe deployed with replication factor at least 3, as we can always tolerate one node going down.

So we've had clusters running for months by now with no OOM, but I assume our environment is quite different.

With the prototype in this issue, I had a load test running for a bit over a week before I killed it, and it never restarted. I can gladly restart it and we can run it longer.
